cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java

```diff
@@ -409,17 +409,24 @@ private void saturatePool() {
 
       // Loop until we either run out of threads to use, or agents (which are keys) to run.
       while (!keys.isEmpty() && runningAgents.map(Semaphore::tryAcquire).orElse(true)) {
-        String agent = keys.remove(0);
-
-        AgentWorker worker = agents.get(agent);
-        ScoreTuple score;
-        if (worker != null && (score = acquireAgent(worker.agent)) != null) {
-          // This score is used to determine if the worker thread running the agent is allowed to
-          // store its results.
-          // If on release of this agent, the scores don't match, this agent was rescheduled by a
-          // separate thread.
-          worker.setScore(score.acquireScore);
-          workers.add(worker);
+        try {
+          String agent = keys.remove(0);
+
+          AgentWorker worker = agents.get(agent);
+          ScoreTuple score;
+          if (worker != null && (score = acquireAgent(worker.agent)) != null) {
+            // This score is used to determine if the worker thread running the agent is allowed to
+            // store its results.
+            // If on release of this agent, the scores don't match, this agent was rescheduled by a
+            // separate thread.
+            worker.setScore(score.acquireScore);
+            workers.add(worker);
+          } else {
+            runningAgents.ifPresent(Semaphore::release);
+          }
+        } catch (Throwable t) {
+          log.error("Failed to add AgentWorker", t);
+          runningAgents.ifPresent(Semaphore::release);
```

Comment:
Bettter to re-throw t to retain the current behavior?  I don't know enough about this code to know which way is better.

Reply:
Re-throw the Throwable will also cause runningAgents to not release permission to the Semaphore.
<img width="1282" alt="image" src="https://user-images.githubusercontent.com/40934357/225497290-937aaab7-373a-4630-a0a6-6419bc2fc6be.png">

Reply:
I figured the additional throw would come after the release you added here, but maybe that still doesn't work.

Reply:
Yes, you're right. Before this PR, it released the acquired permit to the Semaphore [here](https://github.com/spinnaker/clouddriver/blob/909b6ff11dad82268a0b4401f70cabc9b4a3f3ed/cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java#L471), which will run in the worker thread managed by [agentWorkPool](https://github.com/spinnaker/clouddriver/blob/909b6ff11dad82268a0b4401f70cabc9b4a3f3ed/cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java#L427). In this way, before the agentWorker is submitted to the agentWorkPool, the acquired permit will not be released due to an exception throw. Regarding the question of whether to re-throw, I think it is better to ignore, so that the agentWorker can be submitted to the agentWorkPool as much as possible.